src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/SafeEvpPKeyHandle.OpenSsl.cs

```diff
@@ -31,13 +38,53 @@ public SafeEvpPKeyHandle(IntPtr handle, bool ownsHandle)
         {
         }
 
+        internal SafeEvpPKeyHandle(IntPtr handle, IntPtr extraHandle)
+            : base(handle, ownsHandle: true)
+        {
+            ExtraHandle = extraHandle;
+        }
+
         protected override bool ReleaseHandle()
         {
-            Interop.Crypto.EvpPkeyDestroy(handle);
+            Interop.Crypto.EvpPkeyDestroy(handle, ExtraHandle);
+            ExtraHandle = IntPtr.Zero;
+
             SetHandle(IntPtr.Zero);
             return true;
         }
 
+        internal static SafeEvpPKeyHandle GenerateRSAKey(int keySize)
+        {
+            return Interop.Crypto.RsaGenerateKey(keySize);
+        }
+
+        internal static SafeEvpPKeyHandle GenerateECKey(int keySize)
+        {
+            SafeEvpPKeyHandle ret = GenerateECKeyCore(new ECOpenSsl(keySize), out int createdKeySize);
+            Debug.Assert(keySize == createdKeySize);
+            return ret;
+        }
+
+        internal static SafeEvpPKeyHandle GenerateECKey(ECCurve curve, out int keySize)
+        {
+            return  GenerateECKeyCore(new ECOpenSsl(curve), out keySize);
+        }
+
+        internal static SafeEvpPKeyHandle GenerateECKey(ECParameters parameters, out int keySize)
+        {
+            return GenerateECKeyCore(new ECOpenSsl(parameters), out keySize);
+        }
+
+        private static SafeEvpPKeyHandle GenerateECKeyCore(ECOpenSsl ecOpenSsl, out int keySize)
+        {
+            using (ECOpenSsl ec = ecOpenSsl)
+            {
+                SafeEvpPKeyHandle handle = Interop.Crypto.CreateEvpPkeyFromEcKey(ec.Value);
+                keySize = ec.KeySize;
+                return handle;
+            }
+        }
+
```

Comment:
These feel like they belong more in Interop than on this type.  Why are they here?

Reply:
Those construct the SafeEvpPKeyHandle therefore felt like a right place - some of them do more than just interop (ECOpenSsl) and I didn't want to put half of them in the Interop.* files and other half elsewhere. I initially put them in RSA* and ECDsa* classes because that code was extracted from there but realized I can re-use the code between ECDsa and ECDH so it felt like it might be a good idea to put all key constructions in the single place and therefore the choice

Reply:
>  re-use the code between ECDsa and ECDH

That's what the ECOpenSsl class is for.